Sources/NIOHTTPClient/HTTPHandler.swift

```diff
@@ -19,24 +19,40 @@ import NIOHTTP1
 import NIOSSL
 
 public extension HTTPClient {
-    enum Body: Equatable {
-        case byteBuffer(ByteBuffer)
-        case data(Data)
-        case string(String)
-
-        var length: Int {
-            switch self {
-            case .byteBuffer(let buffer):
-                return buffer.readableBytes
-            case .data(let data):
-                return data.count
-            case .string(let string):
-                return string.utf8.count
+    struct Body {
+        typealias ChunkProvider = (@escaping (IOData) -> EventLoopFuture<Void>) -> EventLoopFuture<Void>
+
+        var length: Int?
+        var provider: ChunkProvider
+
+        static func byteBuffer(_ buffer: ByteBuffer) -> Body {
+            return Body(length: buffer.readableBytes) { writer in
+                writer(.byteBuffer(buffer))
+            }
+        }
+
+        static func stream(length: Int? = nil, _ provider: @escaping ChunkProvider) -> Body {
+            return Body(length: length, provider: provider)
+        }
+
+        static func data(_ data: Data) -> Body {
+            return Body(length: data.count) { writer in
+                var buffer = ByteBufferAllocator().buffer(capacity: data.count)
+                buffer.writeBytes(data)
+                return writer(.byteBuffer(buffer))
+            }
+        }
+
+        static func string(_ string: String) -> Body {
+            return Body(length: string.utf8.count) { writer in
+                var buffer = ByteBufferAllocator().buffer(capacity: string.utf8.count)
+                buffer.writeString(string)
```

Comment:
Same - it seems a shame we have to copy the bytes out of the string here.

Reply:
copies are _fast_ (unless you're copying megabytes and more). The allocation here is more troublesome and that could  be avoided.

Generally, if someone needs the maximum performance, they can always prepare a `ByteBuffer` straight away.